### PR TITLE
Embed RPC port to etcd.

### PR DESCRIPTION
**Pull request by huachaohuang**
Remove addr and advertise-addr.

### CODE REVIEW

**Comment by siddontang** [comment]
How does TiKV do? 
Now TiKV can't work.

**Comment by huachaohuang** [comment]
I'm working on it. You can review this first.

File: server/server.go [hunk 1]
```diff
@@ -199,33 +188,49 @@ func (s *Server) isClosed() bool {
...
 
-		c, err := newConn(s, conn)
-		if err != nil {
-			log.Warn(err)
-			conn.Close()
-			continue
-		}
+	conn, bufrw, err := hj.Hijack()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (siddontang) [inline server/server.go]: should we close it directly?
TiKV can't handle this response.

File: pd-client/rpc_worker.go [hunk 2]
```diff
@@ -341,3 +345,45 @@ func (w *rpcWorker) checkResponse(resp *pdpb.Response) error {
...
+		return nil, errors.Trace(err)
+	}
+
+	urls, err := parseUrls(addr)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, url := range urls {
+		conn, err := net.DialTimeout("tcp", url.Host, connectPDTimeout)
```
Comment (qiuyesuifeng) [inline pd-client/rpc_worker.go]: If one url can not be connected, should we try another one instead of return error?

File: server/server.go (hunk 1 above)
Comment (qiuyesuifeng) [inline server/server.go]: Any need to let TiKV log this http response error information?

File: server/server.go [hunk 3]
```diff
@@ -199,33 +188,49 @@ func (s *Server) isClosed() bool {
...
-		s.wg.Add(1)
-		go c.run()
+	err = conn.SetDeadline(time.Time{})
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
 	}
 
-	return nil
+	c, err := newConn(s, conn, bufrw)
```
Comment (qiuyesuifeng) [inline server/server.go]: Can we set readBufferSize/writeBufferSize to bufrw?

File: server/server.go (hunk 1 above)
Comment (huachaohuang) [inline server/server.go]: I think I should just log it in PD and close the connection.

File: server/server.go (hunk 3 above)
Comment (huachaohuang) [inline server/server.go]: I can't find a way to do that, any suggestion?

Comment (qiuyesuifeng) [inline server/server.go]: @siddontang  any suggestion?

File: server/conn.go [hunk 4]
```diff
@@ -38,18 +33,18 @@ type conn struct {
...
 	if !s.isLeader() {
-		return nil, errors.Errorf("server <%s> is not leader, cannot create new connection <%s>", s.cfg.AdvertiseAddr, netConn.RemoteAddr())
+		return nil, errors.Errorf("server <%s> is not leader, cannot create new connection <%s>", s.Name(), netConn.RemoteAddr())
 	}
 
 	c := &conn{
 		s:    s,
-		rb:   bufio.NewReaderSize(netConn, readBufferSize),
-		wb:   bufio.NewWriterSize(netConn, writeBufferSize),
+		rb:   bufrw.Reader,
```
Comment (siddontang) [inline server/conn.go]:
```
rb: bufio.NewReaderSize(bufrw.Reader, readBufferSize),
wb: bufio.NewWriterSize(bufrw.Reader, writeBufferSize),
```

Comment (huachaohuang) [inline server/conn.go]: Will it use two buffer? One in `bufrw.Reader`, another in `rb`?

Comment (siddontang) [inline server/conn.go]: yes, but I have no way but this too. 
The default buf io size is 4k, a little small.

Comment (huachaohuang) [inline server/conn.go]: Seems we can't do that.
When `wb.Flush()`, it only flush to the buffer of `bufrw.Writer`, but it will not flush to the connection.

Comment (siddontang) [inline server/conn.go]: any problem? see https://golang.org/pkg/net/http/#Hijacker
it also uses Flush.

Comment (huachaohuang) [inline server/conn.go]: Yep, then we need to keep both `wb` and `bufrw.Writer`, and we need to flush them both.
I don't think that's a good idea.

File: server/conn.go [hunk 5]
```diff
@@ -38,17 +38,17 @@ type conn struct {
...
 	if !s.isLeader() {
-		return nil, errors.Errorf("server <%s> is not leader, cannot create new connection <%s>", s.cfg.AdvertiseAddr, netConn.RemoteAddr())
+		return nil, errors.Errorf("server <%s> is not leader, cannot create new connection <%s>", s.Name(), netConn.RemoteAddr())
 	}
 
 	c := &conn{
 		s:    s,
-		rb:   bufio.NewReaderSize(netConn, readBufferSize),
+		rb:   bufio.NewReaderSize(bufrw.Reader, readBufferSize),
 		wb:   bufio.NewWriterSize(netConn, writeBufferSize),
```
Comment (siddontang) [inline server/conn.go]: why not using bufrw.Writer here?

File: server/conn.go (hunk 4 above)
Comment (siddontang) [inline server/conn.go]: The same problem for read too, we first read data with bufrw.Reader, then, rb.

Comment (huachaohuang) [inline server/conn.go]: Seems we don't have any better solutions yet, I just use the original `bufrw` here.

File: pd-client/rpc_worker.go [hunk 6]
```diff
@@ -341,3 +345,45 @@ func (w *rpcWorker) checkResponse(resp *pdpb.Response) error {
...
+		if err != nil {
+			continue
+		}
+		err = req.Write(conn)
+		if err != nil {
+			conn.Close()
+			return nil, errors.Trace(err)
```
Comment (siddontang) [inline pd-client/rpc_worker.go]: retry another conn too.

File: server/conn.go [hunk 7]
```diff
@@ -38,17 +38,17 @@ type conn struct {
...
+func newConn(s *Server, netConn net.Conn, bufrw *bufio.ReadWriter) (*conn, error) {
 	s.connsLock.Lock()
 	defer s.connsLock.Unlock()
 
 	if !s.isLeader() {
-		return nil, errors.Errorf("server <%s> is not leader, cannot create new connection <%s>", s.cfg.AdvertiseAddr, netConn.RemoteAddr())
+		return nil, errors.Errorf("server <%s> is not leader, cannot create new connection <%s>", s.Name(), netConn.RemoteAddr())
 	}
 
 	c := &conn{
```
Comment (siddontang) [inline server/conn.go]: I think we need set inner conn read/write buffer size too.

Comment (huachaohuang) [inline server/conn.go]: That will be ugly, I need to do it like this:
```
if muxConn, ok := netConn.(*cmux.MuxConn); ok {
	if tcpConn, ok := muxConn.Conn.(*net.TCPConn); ok {
		tcpConn.SetReadBuffer(readBufferSize)
		tcpConn.SetWriteBuffer(writeBufferSize)
	}
}
```

Comment (siddontang) [inline server/conn.go]: oh, got it.

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**